File: codec/lib/codec.cpp
//
//  codec.cpp
//  codec
//
//  Created by Mr.Panda on 2024/2/14.
//

#include "./codec.h"

static const char* DefaultVideoDecoder = "h264";
static CodecDesc VideoDecoders[] = {
    {"d3d11va", AV_HWDEVICE_TYPE_D3D11VA},
    {"h264_qsv", AV_HWDEVICE_TYPE_QSV},
    {"h264_cuvid", AV_HWDEVICE_TYPE_CUDA},
};

static const char* DefaultVideoEncoder = "libx264";
static CodecDesc VideoEncoders[] = {
    {"h264_qsv", AV_HWDEVICE_TYPE_QSV},
    {"h264_nvenc", AV_HWDEVICE_TYPE_CUDA},
};

#ifdef WIN32
std::optional<CodecContext> create_video_context(CodecKind kind, 
												 std::string& codec, 
												 int width,
												 int height,
												 ID3D11Device* d3d11_device, 
												 ID3D11DeviceContext* d3d11_device_context)
#else
std::optional<CodecContext> create_video_context(CodecKind kind, std::string& codec)
#endif // WIN32
{
    CodecContext ctx;

    if (kind == CodecKind::Encoder)
    {
        ctx.codec = avcodec_find_encoder_by_name(codec.c_str());
    }
    else
    {
        ctx.codec = codec == "d3d11va" ?
            avcodec_find_decoder(AV_CODEC_ID_H264) :
            avcodec_find_decoder_by_name(codec.c_str());
    }

    if (ctx.codec == nullptr)
    {
        return std::nullopt;
    }

    ctx.context = avcodec_alloc_context3(ctx.codec);
    if (ctx.context == nullptr)
    {
        return std::nullopt;
    }

#ifdef WIN32
    if (codec == "d3d11va" || codec == "h264_qsv")
    {
        AVBufferRef* hw_device_ctx = av_hwdevice_ctx_alloc(AV_HWDEVICE_TYPE_D3D11VA);
        if (hw_device_ctx == nullptr)
        {
            avcodec_free_context(&ctx.context);
            return std::nullopt;
        }

        if (codec == "h264_qsv")
        {
            ID3D11Multithread* multithread;
            auto ret = d3d11_device->QueryInterface(__uuidof(ID3D11Multithread),
                                                    (void**)&multithread);
            if (FAILED(ret))
            {
                avcodec_free_context(&ctx.context);
                return std::nullopt;
            }

            multithread->SetMultithreadProtected(TRUE);
            multithread->Release();
        }

        AVHWDeviceContext* hwctx = (AVHWDeviceContext*)hw_device_ctx->data;
        AVD3D11VADeviceContext* d3d11_hwctx = (AVD3D11VADeviceContext*)hwctx->hwctx;
        d3d11_hwctx->device_context = d3d11_device_context;
        d3d11_hwctx->device = d3d11_device;

        if (av_hwdevice_ctx_init(hw_device_ctx) != 0)
        {
            avcodec_free_context(&ctx.context);
            return std::nullopt;
        }

        if (codec == "h264_qsv")
        {
            AVBufferRef* qsv_device_ctx = nullptr;
            if (av_hwdevice_ctx_create_derived(&qsv_device_ctx,
                                               AV_HWDEVICE_TYPE_QSV,
                                               hw_device_ctx,
                                               0) != 0)
            {
                avcodec_free_context(&ctx.context);
                return std::nullopt;
            }

            ctx.context->hw_device_ctx = av_buffer_ref(qsv_device_ctx);

            if (kind == CodecKind::Encoder)
            {
                AVBufferRef* hw_frames_ctx = av_hwframe_ctx_alloc(ctx.context->hw_device_ctx);
                if (!hw_frames_ctx)
                {
                    avcodec_free_context(&ctx.context);
                    return std::nullopt;
                }

                AVHWFramesContext* frames_ctx = (AVHWFramesContext*)(hw_frames_ctx->data);
                frames_ctx->sw_format = AV_PIX_FMT_NV12;
                frames_ctx->format = AV_PIX_FMT_QSV;
                frames_ctx->initial_pool_size = 20;
                frames_ctx->width = width;
                frames_ctx->height = height;

                if (av_hwframe_ctx_init(hw_frames_ctx) < 0)
                {
                    avcodec_free_context(&ctx.context);
                    return std::nullopt;
                }

                ctx.context->hw_frames_ctx = av_buffer_ref(hw_frames_ctx);
            }
        }
        else
        {
            ctx.context->hw_device_ctx = av_buffer_ref(hw_device_ctx);
        }
    }
#endif // WIN32

    return ctx;
}

AVFrame* create_video_frame(AVCodecContext* context)
{
    std::string codec_name = std::string(context->codec->name);
    AVFrame* frame = av_frame_alloc();
    if (frame == nullptr)
    {
        return nullptr;
    }

    frame->width = context->width;
    frame->height = context->height;
    frame->format = context->pix_fmt;

    if (codec_name == "h264_qsv")
    {
        if (av_hwframe_get_buffer(context->hw_frames_ctx, frame, 0) < 0)
        {
            av_frame_free(&frame);
            return nullptr;
        }
    }
    else
    {
        if (av_frame_get_buffer(frame, 0) < 0)
        {
            av_frame_free(&frame);
            return nullptr;
        }
    }

    return frame;
}

template <size_t S>
const char* find_video_codec(CodecDesc(&codecs)[S], CodecKind kind)
{
    AVBufferRef* ctx = nullptr;
    for (auto codec : codecs)
    {
        if (av_hwdevice_ctx_create(&ctx, codec.type, nullptr, nullptr, 0) == 0)
        {
            av_buffer_unref(&ctx);
            return codec.name;
        }
    }

    if (ctx != nullptr)
    {
        av_buffer_unref(&ctx);
    }

    return kind == CodecKind::Encoder ? DefaultVideoEncoder : DefaultVideoDecoder;
}

const char* codec_find_video_encoder()
{
    return find_video_codec(VideoEncoders, CodecKind::Encoder);
}

const char* codec_find_video_decoder()
{
    return find_video_codec(VideoDecoders, CodecKind::Decoder);
}

static Logger GLOBAL_LOGGER = nullptr;

void logger_proc(void* _, int level, const char* message, va_list args)
{
    if (GLOBAL_LOGGER != nullptr && level <= AV_LOG_VERBOSE)
    {
        char str[8192];
        vsnprintf(str, sizeof(str), message, args);
        GLOBAL_LOGGER(level, str);
    }
}

void codec_set_logger(Logger logger)
{
    if (GLOBAL_LOGGER == nullptr)
    {
        GLOBAL_LOGGER = logger;
        av_log_set_callback(logger_proc);
    }
}

void codec_remove_logger()
{
    GLOBAL_LOGGER = nullptr;
}

File: codec/src/codec.rs
use std::str::FromStr;

use mirror_common::{c_str, ffmpeg::*};
use thiserror::Error;

#[cfg(any(target_os = "windows", target_os = "macos"))]
use mirror_common::Size;

#[cfg(target_os = "windows")]
use mirror_common::win32::{windows::core::Interface, Direct3DDevice};

#[derive(Error, Debug)]
pub enum CreateVideoContextError {
    #[error("not found av codec")]
    NotFoundAVCodec,
    #[error("failed to alloc av context")]
    AllocAVContextError,
    #[error("failed to alloc av hardware device context")]
    AllocAVHardwareDeviceContextError,
    #[error("missing direct3d device")]
    MissingDirect3DDevice,
    #[cfg(target_os = "windows")]
    #[error(transparent)]
    SetMultithreadProtectedError(#[from] mirror_common::win32::windows::core::Error),
    #[error("failed to init av hardware device context")]
    InitAVHardwareDeviceContextError,
    #[error("failed to init qsv device context")]
    InitQsvDeviceContextError,
    #[error("failed to alloc av hardware frame context")]
    AllocAVHardwareFrameContextError,
    #[error("failed to init av hardware frame context")]
    InitAVHardwareFrameContextError,
}

#[derive(Debug, Error, Clone, Copy)]
pub enum CodecError {
    #[error("unsupported codecs")]
    NotSupportCodec,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VideoDecoderType {
    H264,
    D3D11,
    Qsv,
    Cuda,
    VideoToolBox,
}

impl ToString for VideoDecoderType {
    fn to_string(&self) -> String {
        match self {
            Self::H264 => "h264",
            Self::D3D11 => "d3d11va",
            Self::Qsv => "h264_qsv",
            Self::Cuda => "h264_cuvid",
            Self::VideoToolBox => "h264_videotoolbox",
        }
        .to_string()
    }
}

impl FromStr for VideoDecoderType {
    type Err = CodecError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Ok(match value {
            "h264" => Self::H264,
            "d3d11va" => Self::D3D11,
            "h264_qsv" => Self::Qsv,
            "h264_cuvid" => Self::Cuda,
            "h264_videotoolbox" => Self::VideoToolBox,
            _ => return Err(CodecError::NotSupportCodec),
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VideoEncoderType {
    X264,
    Qsv,
    Cuda,
    VideoToolBox,
}

impl ToString for VideoEncoderType {
    fn to_string(&self) -> String {
        match self {
            Self::X264 => "libx264",
            Self::Qsv => "h264_qsv",
            Self::Cuda => "h264_nvenc",
            Self::VideoToolBox => "h264_videotoolbox",
        }
        .to_string()
    }
}

impl FromStr for VideoEncoderType {
    type Err = CodecError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Ok(match value {
            "libx264" => Self::X264,
            "h264_qsv" => Self::Qsv,
            "h264_cuvid" => Self::Cuda,
            "h264_videotoolbox" => Self::VideoToolBox,
            _ => return Err(CodecError::NotSupportCodec),
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CodecType {
    Encoder(VideoEncoderType),
    Decoder(VideoDecoderType),
}

impl From<VideoEncoderType> for CodecType {
    fn from(value: VideoEncoderType) -> Self {
        Self::Encoder(value)
    }
}

impl From<VideoDecoderType> for CodecType {
    fn from(value: VideoDecoderType) -> Self {
        Self::Decoder(value)
    }
}

impl CodecType {
    pub fn is_supported(&self) -> bool {
        match self {
            CodecType::Encoder(kind) => {
                if cfg!(target_os = "windows") {
                    *kind != VideoEncoderType::VideoToolBox
                } else if cfg!(target_os = "linux") {
                    *kind == VideoEncoderType::X264
                } else {
                    *kind == VideoEncoderType::X264 || *kind == VideoEncoderType::VideoToolBox
                }
            }
            CodecType::Decoder(kind) => {
                if cfg!(target_os = "windows") {
                    *kind != VideoDecoderType::VideoToolBox
                } else if cfg!(target_os = "linux") {
                    *kind == VideoDecoderType::H264
                } else {
                    *kind == VideoDecoderType::H264 || *kind == VideoDecoderType::VideoToolBox
                }
            }
        }
    }

    pub const fn is_encoder(&self) -> bool {
        if let Self::Encoder(_) = self {
            true
        } else {
            false
        }
    }

    pub fn is_qsv(self) -> bool {
        match self {
            CodecType::Encoder(kind) => kind == VideoEncoderType::Qsv,
            CodecType::Decoder(kind) => kind == VideoDecoderType::Qsv,
        }
    }

    pub fn is_hardware(&self) -> bool {
        match self {
            Self::Decoder(codec) => *codec != VideoDecoderType::H264,
            Self::Encoder(codec) => *codec != VideoEncoderType::X264,
        }
    }

    pub unsafe fn find_av_codec(&self) -> *const AVCodec {
        match self {
            Self::Encoder(kind) => avcodec_find_encoder_by_name(c_str!(kind.to_string())),
            Self::Decoder(kind) => {
                if *kind == VideoDecoderType::D3D11 || *kind == VideoDecoderType::VideoToolBox {
                    avcodec_find_decoder(AVCodecID::AV_CODEC_ID_H264)
                } else {
                    avcodec_find_decoder_by_name(c_str!(kind.to_string()))
                }
            }
        }
    }
}

#[cfg(target_os = "windows")]
pub fn create_video_context(
    context: &mut *mut AVCodecContext,
    kind: CodecType,
    size: Option<Size>,
    direct3d: Option<Direct3DDevice>,
) -> Result<*const AVCodec, CreateVideoContextError> {
    // It is not possible to directly find the d3d11va decoder, so special
    // processing is required here. For d3d11va, the hardware context is initialized
    // below.
    let codec = unsafe { kind.find_av_codec() };
    if codec.is_null() {
        return Err(CreateVideoContextError::NotFoundAVCodec);
    }

    *context = unsafe { avcodec_alloc_context3(codec) };
    if context.is_null() {
        return Err(CreateVideoContextError::AllocAVContextError);
    }

    // The hardware codec is used, and the hardware context is initialized here for
    // the hardware codec.
    if kind.is_hardware() {
        let hw_device_ctx =
            unsafe { av_hwdevice_ctx_alloc(AVHWDeviceType::AV_HWDEVICE_TYPE_D3D11VA) };
        if hw_device_ctx.is_null() {
            return Err(CreateVideoContextError::AllocAVHardwareDeviceContextError);
        }

        // Use externally created d3d devices and do not let ffmpeg create d3d devices
        // itself.
        let direct3d = if let Some(direct3d) = direct3d {
            direct3d
        } else {
            return Err(CreateVideoContextError::MissingDirect3DDevice);
        };

        // Special handling is required for qsv, which requires multithreading to be
        // enabled for the d3d device.
        if kind.is_qsv() {
            if let Err(e) = direct3d.set_multithread_protected(true) {
                return Err(CreateVideoContextError::SetMultithreadProtectedError(e));
            }
        }

        let d3d11_hwctx = unsafe {
            let hwctx = (&mut *hw_device_ctx).data as *mut AVHWDeviceContext;
            &mut *((&mut *hwctx).hwctx as *mut AVD3D11VADeviceContext)
        };

        d3d11_hwctx.device = direct3d.device.as_raw() as *mut _;
        d3d11_hwctx.device_context = direct3d.context.as_raw() as *mut _;

        if unsafe { av_hwdevice_ctx_init(hw_device_ctx) } != 0 {
            return Err(CreateVideoContextError::InitAVHardwareDeviceContextError);
        }

        // Creating a qsv device is a little different, the qsv hardware context needs
        // to be derived from the platform's native hardware context.
        let context_mut = unsafe { &mut **context };
        if kind.is_qsv() {
            let mut qsv_device_ctx = std::ptr::null_mut();
            if unsafe {
                av_hwdevice_ctx_create_derived(
                    &mut qsv_device_ctx,
                    AVHWDeviceType::AV_HWDEVICE_TYPE_QSV,
                    hw_device_ctx,
                    0,
                )
            } != 0
            {
                return Err(CreateVideoContextError::InitQsvDeviceContextError);
            }

            unsafe {
                context_mut.hw_device_ctx = av_buffer_ref(qsv_device_ctx);
            }

            // Similarly, the qsv hardware frame also needs to be created and initialized
            // independently.
            if kind.is_encoder() {
                let hw_frames_ctx = unsafe { av_hwframe_ctx_alloc(context_mut.hw_device_ctx) };
                if hw_frames_ctx.is_null() {
                    return Err(CreateVideoContextError::AllocAVHardwareFrameContextError);
                }

                let size = size.expect("encoder needs init hardware frame for size");
                unsafe {
                    let frames_ctx = &mut *((&mut *hw_frames_ctx).data as *mut AVHWFramesContext);
                    frames_ctx.sw_format = AVPixelFormat::AV_PIX_FMT_NV12;
                    frames_ctx.format = AVPixelFormat::AV_PIX_FMT_QSV;
                    frames_ctx.width = size.width as i32;
                    frames_ctx.height = size.height as i32;
                    frames_ctx.initial_pool_size = 5;
                }

                if unsafe { av_hwframe_ctx_init(hw_frames_ctx) } != 0 {
                    return Err(CreateVideoContextError::InitAVHardwareFrameContextError);
                }

                unsafe {
                    context_mut.hw_frames_ctx = av_buffer_ref(hw_frames_ctx);
                }
            }
        } else {
            unsafe {
                context_mut.hw_device_ctx = av_buffer_ref(hw_device_ctx);
            }
        }
    }

    Ok(codec)
}

#[cfg(target_os = "linux")]
pub fn create_video_context(
    context: &mut *mut AVCodecContext,
    kind: CodecType,
) -> Result<*const AVCodec, CreateVideoContextError> {
    let codec = unsafe { kind.find_av_codec() };
    if codec.is_null() {
        return Err(CreateVideoContextError::NotFoundAVCodec);
    }

    *context = unsafe { avcodec_alloc_context3(codec) };
    if context.is_null() {
        return Err(CreateVideoContextError::AllocAVContextError);
    }

    Ok(codec)
}

#[cfg(target_os = "macos")]
pub fn create_video_context(
    context: &mut *mut AVCodecContext,
    kind: CodecType,
    size: Option<Size>,
) -> Result<*const AVCodec, CreateVideoContextError> {
    let codec = unsafe { kind.find_av_codec() };
    if codec.is_null() {
        return Err(CreateVideoContextError::NotFoundAVCodec);
    }

    *context = unsafe { avcodec_alloc_context3(codec) };
    if context.is_null() {
        return Err(CreateVideoContextError::AllocAVContextError);
    }

    if kind.is_hardware() {
        let mut hw_device_ctx = std::ptr::null_mut();
        if unsafe {
            av_hwdevice_ctx_create(
                &mut hw_device_ctx,
                AVHWDeviceType::AV_HWDEVICE_TYPE_VIDEOTOOLBOX,
                std::ptr::null_mut(),
                std::ptr::null_mut(),
                0,
            )
        } != 0
        {
            return Err(CreateVideoContextError::InitAVHardwareDeviceContextError);
        }

        let context_mut = unsafe { &mut **context };
        context_mut.hw_device_ctx = unsafe { av_buffer_ref(hw_device_ctx) };

        if kind.is_encoder() {
            let hw_frames_ctx = unsafe { av_hwframe_ctx_alloc(context_mut.hw_device_ctx) };
            if hw_frames_ctx.is_null() {
                return Err(CreateVideoContextError::AllocAVHardwareFrameContextError);
            }

            let size = size.expect("encoder needs init hardware frame for size");
            unsafe {
                let frames_ctx = &mut *((&mut *hw_frames_ctx).data as *mut AVHWFramesContext);
                frames_ctx.sw_format = AVPixelFormat::AV_PIX_FMT_NV12;
                frames_ctx.format = AVPixelFormat::AV_PIX_FMT_VIDEOTOOLBOX;
                frames_ctx.width = size.width as i32;
                frames_ctx.height = size.height as i32;
                frames_ctx.initial_pool_size = 5;
            }

            if unsafe { av_hwframe_ctx_init(hw_frames_ctx) } != 0 {
                return Err(CreateVideoContextError::InitAVHardwareFrameContextError);
            }

            unsafe {
                context_mut.hw_frames_ctx = av_buffer_ref(hw_frames_ctx);
            }
        }
    }

    Ok(codec)
}

#[derive(Error, Debug)]
pub enum CreateVideoFrameError {
    #[error("failed to alloc av frame")]
    AllocAVFrameError,
    #[error("failed to alloc hardware av frame buffer")]
    AllocHardwareAVFrameBufferError,
    #[error("failed to alloc av frame buffer")]
    AllocAVFrameBufferError,
}

pub fn create_video_frame(
    frame: &mut *mut AVFrame,
    context: *const AVCodecContext,
) -> Result<(), CreateVideoFrameError> {
    *frame = unsafe { av_frame_alloc() };
    if frame.is_null() {
        return Err(CreateVideoFrameError::AllocAVFrameError);
    }

    let context_ref = unsafe { &*context };
    let frame_mut = unsafe { &mut **frame };

    frame_mut.width = context_ref.width;
    frame_mut.height = context_ref.height;
    frame_mut.format = context_ref.pix_fmt as i32;

    // qsv needs to indicate the use of hardware textures, otherwise qsv will return
    // software textures.
    if !context_ref.hw_device_ctx.is_null() {
        if unsafe { av_hwframe_get_buffer(context_ref.hw_frames_ctx, *frame, 0) } != 0 {
            return Err(CreateVideoFrameError::AllocHardwareAVFrameBufferError);
        }
    } else {
        if unsafe { av_frame_get_buffer(*frame, 0) } != 0 {
            return Err(CreateVideoFrameError::AllocAVFrameBufferError);
        }
    }

    Ok(())
}

pub fn set_option(context: &mut AVCodecContext, key: &str, value: i64) {
    unsafe {
        av_opt_set_int(context.priv_data, c_str!(key), value, 0);
    }
}

pub fn set_str_option(context: &mut AVCodecContext, key: &str, value: &str) {
    unsafe {
        av_opt_set(context.priv_data, c_str!(key), c_str!(value), 0);
    }
}
